compiler/damlc/lib/DA/Cli/Visual.hs

```diff
@@ -81,42 +81,49 @@ darToWorld manifest pkg = AST.initWorldSelf pkgs pkg
 tplName :: LF.Template -> T.Text
 tplName LF.Template {..} = head (LF.unTypeConName tplTypeCon)
 
-handleChoiceAndAction :: ChoiceAndAction -> LF.ChoiceName
+handleChoiceAndAction :: ChoiceAndAction -> (LF.ChoiceName, IsConsuming)
 handleChoiceAndAction (ChoiceAndAction tpl choice _)
-    | LF.chcName choice == LF.ChoiceName "Create" = LF.ChoiceName $ tplName tpl <> "_Create"
-    | LF.chcName choice == LF.ChoiceName "Archive" = LF.ChoiceName $ tplName tpl <> "_Archive"
-    | otherwise = LF.chcName choice
+    | LF.chcName choice == LF.ChoiceName "Create" = (LF.ChoiceName $ tplName tpl <> "_Create", False)
+    | LF.chcName choice == LF.ChoiceName "Archive" = (LF.ChoiceName $ tplName tpl <> "_Archive", True)
+    | otherwise = (LF.chcName choice, LF.chcConsuming choice)
 
+type IsConsuming = Bool
 -- Making choiceName is very weird
-handleCreateAndArchive :: TemplateChoiceAction -> [LF.ChoiceName]
+handleCreateAndArchive :: TemplateChoiceAction -> [(LF.ChoiceName, IsConsuming)]
 handleCreateAndArchive TemplateChoiceAction {..} = [createChoice, archiveChoice] ++ map handleChoiceAndAction choiceAndAction
-    where archiveChoice = LF.ChoiceName $ tplName template <> "_Archive"
-          createChoice = LF.ChoiceName $ tplName template <> "_Create"
+    where archiveChoice = (LF.ChoiceName $ tplName template <> "_Archive", True)
+          createChoice = (LF.ChoiceName $ tplName template <> "_Create", False)
 
+data ChoiceStyle = ChoiceStyle
+    { nodeId :: Int
+    , consuming :: Bool
+    }
 -- This is used to generate the node ids and use as look up table
-choiceNameWithId :: [TemplateChoiceAction] -> Map.Map LF.ChoiceName Int
-choiceNameWithId tplChcActions = Map.fromList $ zip choiceActions [0..]
+
+choiceNameWithId :: [TemplateChoiceAction] -> Map.Map LF.ChoiceName ChoiceStyle
+choiceNameWithId tplChcActions = Map.fromList choiceWithIds
   where choiceActions = concatMap handleCreateAndArchive tplChcActions
+        choiceWithIds = map (\ ((cName, consume) , id) -> (cName, ChoiceStyle id consume) ) $ zip choiceActions [0..]
 
-nodeIdForChoice :: Map.Map LF.ChoiceName Int -> LF.ChoiceName -> Int
+nodeIdForChoice :: Map.Map LF.ChoiceName ChoiceStyle -> LF.ChoiceName -> ChoiceStyle
 nodeIdForChoice nodeLookUp chc = case Map.lookup chc nodeLookUp of
   Just node -> node
   Nothing -> error "Template node lookup failed"
 
 -- probably storing the choice is a better Idea, as we can determine what kind of choice it is.
 data SubGraph = SubGraph
-    { nodes :: [(LF.ChoiceName ,Int)]
+    { nodes :: [(LF.ChoiceName, ChoiceStyle)]
     , clusterTemplate :: LF.Template
     }
```

Comment:
I would put all these type definitions at the top of the file, preferably in some kind of logical order.

Reply:
Yup moved these up